### PR TITLE
Implemented the Streaming Feature to stream vectors from Java to JNI layer to enable creation of larger segments for vector indices

### PR DESCRIPTION
**Pull request by navneet1v**
### Description
Implemented the Streaming Feature to stream vectors from Java to JNI layer to enable creation of larger segments for vector indices

Changes include:
1. Add the interface for streaming the vectors from java to jni layer with initial capacity (#1586)
2. Integrating storeVectors interfaces with createIndex and createIndexTemplate functions. (#1588)
3. Update KNN80BinaryDocValues reader count live docs and use live docs as initial capacity to initialize vector address(#1595)
4. Move free vectorAddress from Java to JNI layer to reduce the memory footprint for Nmslib (#1602)

All these changes are already merged in feature/stream-vectors branch of k-NN https://github.com/opensearch-project/k-NN/tree/feature/stream-vectors
 
## JNI Test
```
(base) 16:32 ~/workplace/k-NN/jni (stream-vectors)$ ./bin/jni_test 
Running main() from /Users/navneev/workplace/k-NN/jni/googletest-src/googletest/src/gtest_main.cc
[==========] Running 22 tests from 20 test suites.
[----------] Global test environment set-up.
[----------] 1 test from FaissCreateIndexTest
[ RUN      ] FaissCreateIndexTest.BasicAssertions
[       OK ] FaissCreateIndexTest.BasicAssertions (5 ms)
[----------] 1 test from FaissCreateIndexTest (5 ms total)

[----------] 1 test from FaissCreateIndexFromTemplateTest
[ RUN      ] FaissCreateIndexFromTemplateTest.BasicAssertions
[       OK ] FaissCreateIndexFromTemplateTest.BasicAssertions (3 ms)
[----------] 1 test from FaissCreateIndexFromTemplateTest (3 ms total)

[----------] 3 tests from FaissLoadIndexTest
[ RUN      ] FaissLoadIndexTest.BasicAssertions
[       OK ] FaissLoadIndexTest.BasicAssertions (3 ms)
[ RUN      ] FaissLoadIndexTest.HNSWPQDisableSdcTable
WARNING clustering 256 points to 16 centroids: please provide at least 624 training points
[       OK ] FaissLoadIndexTest.HNSWPQDisableSdcTable (368 ms)
[ RUN      ] FaissLoadIndexTest.IVFPQDisablePrecomputeTable
WARNING clustering 256 points to 16 centroids: please provide at least 624 training points
[       OK ] FaissLoadIndexTest.IVFPQDisablePrecomputeTable (379 ms)
[----------] 3 tests from FaissLoadIndexTest (751 ms total)

[----------] 1 test from FaissQueryIndexTest
[ RUN      ] FaissQueryIndexTest.BasicAssertions
[       OK ] FaissQueryIndexTest.BasicAssertions (4 ms)
[----------] 1 test from FaissQueryIndexTest (4 ms total)

[----------] 1 test from FaissQueryIndexWithFilterTest1435
[ RUN      ] FaissQueryIndexWithFilterTest1435.BasicAssertions
[       OK ] FaissQueryIndexWithFilterTest1435.BasicAssertions (10 ms)
[----------] 1 test from FaissQueryIndexWithFilterTest1435 (10 ms total)

[----------] 1 test from FaissQueryIndexWithParentFilterTest
[ RUN      ] FaissQueryIndexWithParentFilterTest.BasicAssertions
[       OK ] FaissQueryIndexWithParentFilterTest.BasicAssertions (5 ms)
[----------] 1 test from FaissQueryIndexWithParentFilterTest (5 ms total)

[----------] 1 test from FaissFreeTest
[ RUN      ] FaissFreeTest.BasicAssertions
[       OK ] FaissFreeTest.BasicAssertions (0 ms)
[----------] 1 test from FaissFreeTest (0 ms total)

[----------] 1 test from FaissInitLibraryTest
[ RUN      ] FaissInitLibraryTest.BasicAssertions
[       OK ] FaissInitLibraryTest.BasicAssertions (0 ms)
[----------] 1 test from FaissInitLibraryTest (0 ms total)

[----------] 1 test from FaissTrainIndexTest
[ RUN      ] FaissTrainIndexTest.BasicAssertions
[       OK ] FaissTrainIndexTest.BasicAssertions (0 ms)
[----------] 1 test from FaissTrainIndexTest (0 ms total)

[----------] 1 test from FaissCreateHnswSQfp16IndexTest
[ RUN      ] FaissCreateHnswSQfp16IndexTest.BasicAssertions
[       OK ] FaissCreateHnswSQfp16IndexTest.BasicAssertions (4 ms)
[----------] 1 test from FaissCreateHnswSQfp16IndexTest (4 ms total)

[----------] 1 test from FaissIsSharedIndexStateRequired
[ RUN      ] FaissIsSharedIndexStateRequired.BasicAssertions
[       OK ] FaissIsSharedIndexStateRequired.BasicAssertions (0 ms)
[----------] 1 test from FaissIsSharedIndexStateRequired (0 ms total)

[----------] 1 test from FaissInitAndSetSharedIndexState
[ RUN      ] FaissInitAndSetSharedIndexState.BasicAssertions
WARNING clustering 256 points to 16 centroids: please provide at least 624 training points
[       OK ] FaissInitAndSetSharedIndexState.BasicAssertions (353 ms)
[----------] 1 test from FaissInitAndSetSharedIndexState (353 ms total)

[----------] 1 test from IDGrouperBitMapTest
[ RUN      ] IDGrouperBitMapTest.BasicAssertions
[       OK ] IDGrouperBitMapTest.BasicAssertions (0 ms)
[----------] 1 test from IDGrouperBitMapTest (0 ms total)

[----------] 1 test from NmslibIndexWrapperSearchTest
[ RUN      ] NmslibIndexWrapperSearchTest.BasicAssertions
[       OK ] NmslibIndexWrapperSearchTest.BasicAssertions (0 ms)
[----------] 1 test from NmslibIndexWrapperSearchTest (0 ms total)

[----------] 1 test from NmslibCreateIndexTest
[ RUN      ] NmslibCreateIndexTest.BasicAssertions
[       OK ] NmslibCreateIndexTest.BasicAssertions (1 ms)
[----------] 1 test from NmslibCreateIndexTest (1 ms total)

[----------] 1 test from NmslibLoadIndexTest
[ RUN      ] NmslibLoadIndexTest.BasicAssertions
[       OK ] NmslibLoadIndexTest.BasicAssertions (1 ms)
[----------] 1 test from NmslibLoadIndexTest (1 ms total)

[----------] 1 test from NmslibQueryIndexTest
[ RUN      ] NmslibQueryIndexTest.BasicAssertions
[       OK ] NmslibQueryIndexTest.BasicAssertions (2 ms)
[----------] 1 test from NmslibQueryIndexTest (2 ms total)

[----------] 1 test from NmslibFreeTest
[ RUN      ] NmslibFreeTest.BasicAssertions
[       OK ] NmslibFreeTest.BasicAssertions (0 ms)
[----------] 1 test from NmslibFreeTest (0 ms total)

[----------] 1 test from NmslibInitLibraryTest
[ RUN      ] NmslibInitLibraryTest.BasicAssertions
[       OK ] NmslibInitLibraryTest.BasicAssertions (0 ms)
[----------] 1 test from NmslibInitLibraryTest (0 ms total)

[----------] 1 test from CommonsTests
[ RUN      ] CommonsTests.BasicAssertions
[       OK ] CommonsTests.BasicAssertions (0 ms)
[----------] 1 test from CommonsTests (0 ms total)

[----------] Global test environment tear-down
[==========] 22 tests from 20 test suites ran. (1148 ms total)
[  PASSED  ] 22 tests.

```

### Issues Resolved
https://github.com/opensearch-project/k-NN/issues/1506
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by martin-gaievski — APPROVED** [review]

**Comment by navneet1v** [comment]
BWC fix PR: https://github.com/opensearch-project/k-NN/pull/1605 . The failures in BWC is not related to this PR.

**Review by naveentatikonda — APPROVED** [review]

**Comment by naveentatikonda** [comment]
Build is failing on Windows. Can you please check

**Comment by navneet1v** [comment]
> Build is failing on Windows. Can you please check

```
make[3]: *** [CMakeFiles/opensearchknn_nmslib.dir/build.make:120: release/opensearchknn_nmslib.dll] Error 1
make[2]: *** [CMakeFiles/Makefile2:287: CMakeFiles/opensearchknn_nmslib.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:294: CMakeFiles/opensearchknn_nmslib.dir/rule] Error 2
make: *** [Makefile:202: opensearchknn_nmslib] Error 2

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':buildJniLib'.
> Process 'command 'make'' finished with non-zero exit value 2

```

For a long time I have not seen windows CI successful

File: src/main/java/org/opensearch/knn/jni/JNICommons.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+
+/**
+ * Common class for providing the JNI related functionality to various JNIServices.
+ */
+public class JNICommons {
+
+    static {
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/jni/JNICommons.java:24]: Do we need to explicitly ensure this gets called before faiss and nmslib get called?

File: jni/src/nmslib_wrapper.cpp [hunk 2]
```diff
@@ -134,18 +147,18 @@ void knn_jni::nmslib_wrapper::CreateIndex(knn_jni::JNIUtilInterface * jniUtil, J
...
-            if (dim != jniUtil->GetJavaFloatArrayLength(env, floatArrayJ)) {
-                throw std::runtime_error("Dimension of vectors is inconsistent");
-            }
-
-            floatArrayCpp = jniUtil->GetFloatArrayElements(env, floatArrayJ, nullptr);
-            memcpy(ptr, floatArrayCpp, vectorSizeInBytes);
-            jniUtil->ReleaseFloatArrayElements(env, floatArrayJ, floatArrayCpp, JNI_ABORT);
+            memcpy(ptr, &(inputVectors->at(vectorPointer)), vectorSizeInBytes);
             ptr += vectorSizeInBytes;
+            vectorPointer += dim;
```
[the remark this answers is not in the document]
Comment (navneet1v) [reply jni/src/nmslib_wrapper.cpp:152]: Yes thats correct

File: src/main/java/org/opensearch/knn/jni/JNICommons.java (hunk 1 above)
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/jni/JNICommons.java:24]: I didn't find any issue while running the cluster.

Comment (navneet1v) [reply src/main/java/org/opensearch/knn/jni/JNICommons.java:24]: and as of now there is no dependency between native libs and the commons. So we should be good for now.

Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/jni/JNICommons.java:24]: [This](https://github.com/navneet1v/k-NN/blob/stream-vectors/jni/src/org_opensearch_knn_jni_JNICommons.cpp#L21-L37) wont get called if one of the native libraries first calls it correct?

> and as of now there is no dependency between native libs and the commons. So we should be good for now.

I dont understand. https://github.com/navneet1v/k-NN/blob/stream-vectors/jni/include/jni_util.h is depended upon by both faiss and nmslib, right?

Comment (navneet1v) [reply src/main/java/org/opensearch/knn/jni/JNICommons.java:24]: So I validated that if Faiss Library gets loaded first too, then also we don't see any issue. Here are the steps I did.

1. First I added the log messages when Faiss and Commons lib were getting loaded.
2. I created a faiss index and did the indexing. This triggered Commons to get loaded first and then the faiss.
3. After this I killed the cluster and ran the cluster again with `--preserve-data` flag. 
4. After this I ran a search query on the index created with faiss engine. The search was successful. Only Faiss got loaded this time.
5. I ran the indexing again and this time commons got loaded. 
6. Did the search again, the search was successful.

Logs for step 3 and 4

```
[2024-04-10T13:43:09,162][INFO ][stdout                   ] [integTest-0] Loading Faiss library.....
[2024-04-10T13:43:09,174][WARN ][stderr                   ] [integTest-0] SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
[2024-04-10T13:43:09,175][WARN ][stderr                   ] [integTest-0] SLF4J: Defaulting to no-operation (NOP) logger implementation
[2024-04-10T13:43:09,175][WARN ][stderr                   ] [integTest-0] SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
[2024-04-10T13:44:03,171][INFO ][stdout                   ] [integTest-0] Loading JNICommons library.....

```

Comment (navneet1v) [reply src/main/java/org/opensearch/knn/jni/JNICommons.java:24]: From java side I also checked if loadLibrary is called multiple times then there is no issues.

Doc: https://docs.oracle.com/javase/7/docs/api/java/lang/Runtime.html#loadLibrary%28java.lang.String%29:~:text=If%20this%20method%20is%20called%20more%20than%20once%20with%20the%20same%20library%20name%2C%20the%20second%20and%20subsequent%20calls%20are%20ignored.

**Review by jmazanec15 — APPROVED** [review]

**Pull request merged**